libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,747 @@
+    if( packetId == MQTT_PACKET_ID_INVALID || qos == MQTTQoS0 )
+    {
+        status = MQTTBadParameter;
+    }
+    else
+    {
+        /* Start from end so first available index will be populated. */
+        for( index = recordLen - 1; index >= 0; index-- )
+        {
+            if( records[ index ].packetId == MQTT_PACKET_ID_INVALID )
+            {
+                availableIndex = index;
+            }
+            else if( records[ index ].packetId == packetId )
+            {
+                /* Collision. */
+                status = MQTTStateCollision;
+                availableIndex = recordLen;
+                break;
+            }
+            else
+            {
+                /* Empty else clause. */
+            }   
+        }
+    }
+
+    if( availableIndex < recordLen )
+    {
+        records[ availableIndex ].packetId = packetId;
+        records[ availableIndex ].qos = qos;
+        records[ availableIndex ].publishState = publishState;
+        status = MQTTSuccess;
+    }
+
+    return status;
+}
+
+static MQTTStatus_t _updateRecord( MQTTPubAckInfo_t * records,
+                                   size_t recordLen,
+                                   size_t recordIndex,
+                                   uint16_t packetId,
+                                   MQTTQoS_t qos,
+                                   MQTTPublishState_t newState,
+                                   bool shouldStore,
+                                   bool shouldDelete )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    if( shouldStore )
+    {
+        status = _addRecord( records,
+                             recordLen,
+                             packetId,
+                             qos,
+                             newState );
+    }
+    else if( recordIndex >= recordLen )
+    {
+        /* Something's wrong because the record should exist. */
+        status = MQTTBadParameter;
+    }
+    else if( shouldDelete )
+    {
+        status = _deleteRecord( records,
+                                recordLen,
+                                recordIndex );
+    }
+    else
+    {
+        /* Sanity checks. */
+        if( records[ recordIndex ].packetId != packetId )
+        {
+            status = MQTTBadParameter;
+        }
+        else if( records[ recordIndex ].qos != qos )
+        {
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            records[ recordIndex ].publishState = newState;
+        }
+    }
+    
+    return status;
+}
+
+MQTTPublishState_t _MQTT_CalculateState( MQTTPublishType_t packetType,
+                                         MQTTStateOperation_t opType,
+                                         MQTTQoS_t qos,
+                                         MQTTPacketIDState_t idStatus )
+{
+    MQTTPublishState_t calculatedState = MQTTInvalidState;
+    switch( qos )
+    {
+        case MQTTQoS0:
+            if( idStatus == ID_UNKNOWN && packetType == MQTT_PUBLISH )
```

Comment:
Add parentheses around the sub-expressions in this compound statement.

Reply:
Fixed